File: 01-OPERATING-SYSTEM/api_gateway/middleware/logging_middleware.py
"""
Logging Middleware - وسيط التسجيل
Request/Response logging middleware
"""

import logging
import time
from typing import Callable
from fastapi import Request
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import Response

logger = logging.getLogger(__name__)


class LoggingMiddleware(BaseHTTPMiddleware):
    """
    Logging Middleware
    Logs all HTTP requests and responses
    """
    
    async def dispatch(self, request: Request, call_next: Callable) -> Response:
        """
        Process request and log information
        
        Args:
            request: HTTP request
            call_next: Next middleware/handler
            
        Returns:
            HTTP response
        """
        # Start timer
        start_time = time.time()
        
        # Log request
        client_ip = request.client.host if request.client else "unknown"
        method = request.method
        path = request.url.path
        query_params = str(request.query_params) if request.query_params else ""
        
        logger.info(
            f"📥 {method} {path}{'?' + query_params if query_params else ''} "
            f"from {client_ip}"
        )
        
        # Process request
        try:
            response = await call_next(request)
            
            # Calculate duration
            duration = time.time() - start_time
            
            # Log response
            status_code = response.status_code
            status_emoji = "✅" if 200 <= status_code < 300 else "⚠️" if 300 <= status_code < 400 else "❌"
            
            logger.info(
                f"{status_emoji} {method} {path} → {status_code} "
                f"({duration:.3f}s)"
            )
            
            # Add custom headers
            response.headers["X-Process-Time"] = f"{duration:.3f}"
            
            return response
            
        except Exception as e:
            # Log error
            duration = time.time() - start_time
            logger.error(
                f"❌ {method} {path} → ERROR: {str(e)} ({duration:.3f}s)",
                exc_info=True
            )
            raise

File: database-core/src/presentation/api/middleware/logging.middleware.ts
/**
 * Logging Middleware - Middleware لتسجيل الطلبات
 *
 * Middleware لتسجيل جميع الطلبات الواردة
 */

import { Request, Response, NextFunction } from 'express'
import { logger } from '../../../shared/utils/logger'

/**
 * Logging Middleware
 */
export function loggingMiddleware(req: Request, res: Response, next: NextFunction): void {
  const startTime = Date.now()

  // استخدام originalUrl للحصول على المسار الكامل
  const endpoint = req.originalUrl || req.url || req.path

  // تسجيل الطلب الوارد
  logger.info('Incoming request', {
    method: req.method,
    path: endpoint,
    query: req.query,
    ip: req.ip,
    userAgent: req.get('user-agent'),
  })

  // تسجيل الاستجابة عند الانتهاء
  res.on('finish', () => {
    const duration = Date.now() - startTime
    logger.info('Request completed', {
      method: req.method,
      path: endpoint,
      statusCode: res.statusCode,
      duration,
    })
  })

  next()
}
